src/libsync/foldermetadata.h

```diff
@@ -0,0 +1,247 @@
+
+    [[nodiscard]] QVector<EncryptedFile> files() const;
+
+    [[nodiscard]] bool isValid() const;
+
+    [[nodiscard]] bool isFileDropPresent() const;
+
+    [[nodiscard]] bool isRootEncryptedFolder() const;
+
+    [[nodiscard]] bool encryptedMetadataNeedUpdate() const;
+
+    [[nodiscard]] bool moveFromFileDropToFiles();
+
+    // adds a user to have access to this folder (always generates new metadata key)
+    [[nodiscard]] bool addUser(const QString &userId, const QSslCertificate &certificate);
+    // removes a user from this folder and removes and generates a new metadata key
+    [[nodiscard]] bool removeUser(const QString &userId);
+
+    [[nodiscard]] const QByteArray metadataKeyForEncryption() const;
+    [[nodiscard]] const QByteArray metadataKeyForDecryption() const;
+    [[nodiscard]] const QSet<QByteArray> &keyChecksums() const;
+
+    [[nodiscard]] QByteArray encryptedMetadata();
+
+    [[nodiscard]] EncryptionStatusEnums::ItemEncryptionStatus existingMetadataEncryptionStatus() const;
+    [[nodiscard]] EncryptionStatusEnums::ItemEncryptionStatus encryptedMetadataEncryptionStatus() const;
+
+    [[nodiscard]] bool isVersion2AndUp() const;
+
+    [[nodiscard]] quint64 newCounter() const;
+
+    [[nodiscard]] QByteArray metadataSignature() const;
+
+    [[nodiscard]] QByteArray initialMetadata() const;
+
+public slots:
+    void addEncryptedFile(const EncryptedFile &f);
+    void removeEncryptedFile(const EncryptedFile &f);
+    void removeAllEncryptedFiles();
+
+private:
+    [[nodiscard]] QByteArray encryptedMetadataLegacy();
+
+    [[nodiscard]] bool verifyMetadataKey(const QByteArray &metadataKey) const;
+
+    [[nodiscard]] QByteArray encryptDataWithPublicKey(const QByteArray &data, const QSslKey key) const;
+    [[nodiscard]] QByteArray decryptDataWithPrivateKey(const QByteArray &data) const;
+
+    [[nodiscard]] QByteArray encryptJsonObject(const QByteArray& obj, const QByteArray pass) const;
+    [[nodiscard]] QByteArray decryptJsonObject(const QByteArray& encryptedJsonBlob, const QByteArray& pass) const;
+
+    [[nodiscard]] bool checkMetadataKeyChecksum(const QByteArray &metadataKey, const QByteArray &metadataKeyChecksum) const;
+
+    [[nodiscard]] QByteArray computeMetadataKeyChecksum(const QByteArray &metadataKey) const;
+
+    [[nodiscard]] EncryptedFile parseEncryptedFileFromJson(const QString &encryptedFilename, const QJsonValue &fileJSON) const;
+
+    [[nodiscard]] QJsonObject convertFileToJsonObject(const EncryptedFile *encryptedFile) const;
+
+    [[nodiscard]] MetadataVersion latestSupportedMetadataVersion() const;
+
+    [[nodiscard]] bool parseFileDropPart(const QJsonDocument &doc);
+
+    void setFileDrop(const QJsonObject &fileDrop);
+
+    static EncryptionStatusEnums::ItemEncryptionStatus fromMedataVersionToItemEncryptionStatus(const MetadataVersion &metadataVersion);
+    static MetadataVersion fromItemEncryptionStatusToMedataVersion(const EncryptionStatusEnums::ItemEncryptionStatus &encryptionStatus);
```

Comment:
for enum values, please do not use `const type&`, this is slower than passing by value